File: apps/mobile/lib/data/eta/mock_eta_provider.dart
import '../../core/utils/geo.dart';
import '../models/eta_result.dart';
import 'eta_provider.dart';

/// Harici API'ye bağımlı olmayan basit ETA hesaplayıcı.
///
/// Kalan mesafeyi durak zinciri üzerinden toplar, ortalama hız ve durak duruş
/// sürelerini ekleyerek tahmini varış süresini üretir.
class MockEtaProvider implements EtaProvider {
  @override
  String get name => 'mock';

  @override
  EtaResult calculate(EtaInput input) {
    final now = input.now ?? DateTime.now();
    final target = input.targetStopIndex
        .clamp(input.nextStopIndex, input.stops.length - 1);

    // 1) Kalan mesafe: araç -> sıradaki durak -> ... -> hedef durak
    var distance = 0.0;
    if (input.nextStopIndex <= target) {
      distance += haversineMeters(input.vehicleLocation, input.stops[input.nextStopIndex].point);
      for (var i = input.nextStopIndex; i < target; i++) {
        distance += haversineMeters(input.stops[i].point, input.stops[i + 1].point);
      }
    }

    final remainingStops = (target - input.nextStopIndex + 1).clamp(0, input.stops.length);

    // 2) Sürüş süresi (saniye)
    final speedMs = (input.averageSpeedKmh * 1000 / 3600).clamp(1, double.infinity);
    final driveSeconds = distance / speedMs;

    // 3) Ara duraklardaki duruş süreleri (hedef durak hariç)
    final dwellSeconds = ((remainingStops - 1).clamp(0, 1 << 31)) * input.dwellSecondsPerStop;

    final etaSeconds = driveSeconds + dwellSeconds;
    final etaMinutes = (etaSeconds / 60).round().clamp(0, 1 << 31);

    final estimated = now.add(Duration(seconds: etaSeconds.round()));
    final delayMinutes =
        (estimated.difference(input.plannedArrivalAt).inSeconds / 60).round().clamp(0, 1 << 31);

    return EtaResult(
      remainingStops: remainingStops,
      remainingDistanceMeters: distance.roundToDouble(),
      etaMinutes: etaMinutes,
      plannedArrivalAt: input.plannedArrivalAt,
      estimatedArrivalAt: estimated,
      delayMinutes: delayMinutes,
    );
  }
}

File: apps/mobile-legacy-rn/src/services/eta/MockETAProvider.ts
import type { EtaResult } from '@servis/shared';
import { haversineMeters } from '@/lib/geo';
import type { ETAProvider, EtaInput } from './types';

/**
 * Harici API'ye bağımlı olmayan basit ETA hesaplayıcı.
 * Kalan mesafeyi durak zinciri üzerinden toplar, ortalama hız ve
 * durak duruş sürelerini ekleyerek tahmini varış süresini üretir.
 */
export class MockETAProvider implements ETAProvider {
  readonly name = 'mock';

  async calculate(input: EtaInput): Promise<EtaResult> {
    const {
      vehicleLocation,
      nextStopIndex,
      targetStopIndex,
      stops,
      averageSpeedKmh,
      dwellSecondsPerStop,
      plannedArrivalAt,
    } = input;

    const now = input.now ? new Date(input.now) : new Date();
    const clampedTarget = Math.min(Math.max(targetStopIndex, nextStopIndex), stops.length - 1);

    // 1) Kalan mesafe: araç -> sıradaki durak -> ... -> hedef durak
    let distance = 0;
    if (nextStopIndex <= clampedTarget) {
      distance += haversineMeters(vehicleLocation, stops[nextStopIndex]!);
      for (let i = nextStopIndex; i < clampedTarget; i++) {
        distance += haversineMeters(stops[i]!, stops[i + 1]!);
      }
    }

    const remainingStops = Math.max(0, clampedTarget - nextStopIndex + 1);

    // 2) Sürüş süresi (saniye) = mesafe / hız
    const speedMs = Math.max(1, (averageSpeedKmh * 1000) / 3600);
    const driveSeconds = distance / speedMs;

    // 3) Ara duraklardaki duruş süreleri (hedef durak hariç)
    const dwellSeconds = Math.max(0, remainingStops - 1) * dwellSecondsPerStop;

    const etaSeconds = driveSeconds + dwellSeconds;
    const etaMinutes = Math.max(0, Math.round(etaSeconds / 60));

    // 4) Varış tahmini ve gecikme
    const estimatedArrival = new Date(now.getTime() + etaSeconds * 1000);
    const planned = new Date(plannedArrivalAt);
    const delayMinutes = Math.max(
      0,
      Math.round((estimatedArrival.getTime() - planned.getTime()) / 60000),
    );

    return {
      remaining_stops: remainingStops,
      remaining_distance_meters: Math.round(distance),
      eta_minutes: etaMinutes,
      planned_arrival_at: planned.toISOString(),
      estimated_arrival_at: estimatedArrival.toISOString(),
      delay_minutes: delayMinutes,
    };
  }
}
